### PR TITLE
Makes various improvements to the CLI

### PR DESCRIPTION
**Pull request by dposada**
- [x] read version from a shared `version.py` file
- [x] on `list`, change `--state ...` to `--all`, `--waiting`, etc
- [x] on `list`, default should be waiting + running, not just running
- [x] on `list` lookback help, `"X"` -> `"LOOKBACK"`
- [x] on `list`, add support for `--submitted-after` and `--submitted-before`
- [x] case insensitive cluster names
- [x] increase the default read timeout
- [x] don't show `KeyboardInterrupt` garbage on `Ctrl+C`
- [x] list all default flag values in help text
- [x] on `submit`, if we timeout, tell the user that their job may have been submitted
- [x] on `wait`, fix the weird formatting issue

### CODE REVIEW

File: cli/cook/subcommands/list.py [hunk 1]
```diff
@@ -70,16 +75,51 @@ def show_data(cluster_job_pairs):
...
     user = args.get('user')
     lookback_hours = args.get('lookback')
+    submitted_after = args.get('submitted_after')
+    submitted_before = args.get('submitted_before')
     name = args.get('name')
     limit = args.get('limit')
 
-    query_result = query(clusters, state, user, lookback_hours, name, limit)
+    if lookback_hours and (submitted_after or submitted_before):
+        raise Exception('You cannot specify both lookback hours and submitted after / before times.')
```
Comment (DaoWen) [inline cli/cook/subcommands/list.py]: This pattern seems to be repeated in a couple of places. As far as I can tell, these checks are really enforcing mutually exclusive command-line arguments. Maybe we should do that at the level of `argparse` instead? 
https://docs.python.org/3/library/argparse.html#mutual-exclusion

Comment (dposada) [reply cli/cook/subcommands/list.py]: The mutual exclusion groups are nice, but they don't seem to support "either this one, or either or both of these other two" (you can either do lookback, or submitted after and/or submitted before).

File: cli/cook/subcommands/list.py [hunk 2]
```diff
@@ -70,16 +75,51 @@ def show_data(cluster_job_pairs):
...
 
 
+def date_time_string_to_ms_since_epoch(date_time_string):
+    """Converts the given date_time_string (e.g. '5 minutes ago') to milliseconds since epoch"""
+    calendar = parsedatetime.Calendar()
+    local_tz = tzlocal.get_localzone()
+    dt, parse_status = calendar.parseDT(datetimeString=date_time_string, tzinfo=local_tz)
+    if parse_status == 1 or parse_status == 2:
+        ms_since_epoch = int((dt - EPOCH).total_seconds() * 1000)
+        logging.debug(f'parsed "{date_time_string}" as {dt} (ms = {ms_since_epoch}) using time zone {local_tz}')
```
[the remark this answers is not in the document]
Comment (dposada) [reply cli/cook/subcommands/list.py]: :+1:

File: cli/cook/subcommands/list.py [hunk 3]
```diff
@@ -70,16 +75,51 @@ def show_data(cluster_job_pairs):
...
+    if lookback_hours and (submitted_after or submitted_before):
+        raise Exception('You cannot specify both lookback hours and submitted after / before times.')
+
+    if submitted_after or submitted_before:
+        start_ms = date_time_string_to_ms_since_epoch(submitted_after or f'{DEFAULT_LOOKBACK_HOURS} hours ago')
+        end_ms = date_time_string_to_ms_since_epoch(submitted_before or 'now')
+    elif lookback_hours:
+        start_ms, end_ms = lookback_hours_to_range(lookback_hours)
+    else:
+        start_ms, end_ms = lookback_hours_to_range(DEFAULT_LOOKBACK_HOURS)
```
[the remark this answers is not in the document]
Comment (dposada) [reply cli/cook/subcommands/list.py]: Good call :+1:

File: cli/cook/util.py [hunk 4]
```diff
@@ -49,7 +54,7 @@ def load_first_json_file(paths=None):
     """Returns the contents of the first parseable JSON file in a list of paths."""
     if paths is None:
```
[the remark this answers is not in the document]
Comment (dposada) [reply cli/cook/util.py]: Good point, fixed

File: integration/tests/cook/cli.py [hunk 5]
```diff
@@ -107,10 +107,10 @@ def __exit__(self, _, __, ___):
         os.remove(self.path)
 
 
-def list_jobs(cook_url=None, list_flags=None):
+def list_jobs(cook_url=None, list_flags=None, flags=None):
     """Invokes the list subcommand"""
-    args = 'list %s' % list_flags + ' ' if list_flags else ''
-    cp = cli(args, cook_url)
+    args = f'list{" " + list_flags if list_flags else ""}'
```
[the remark this answers is not in the document]
Comment (dposada) [reply integration/tests/cook/cli.py]: Me too, fixed

**Comment by dposada** [comment]
@DaoWen this is ready for another round of feedback. Note that in addition to the changes you suggested, I've replaced the use of a library for relative date (e.g. "2 days ago") parsing and rolled my own instead. There's a comment on the `dateparser.parse` function explaining why.

File: cli/cook/subcommands/list.py (hunk 1 above)
Comment (DaoWen) [reply cli/cook/subcommands/list.py]: I think you can nest groups, which would let you do that. But let's just ask this question first: Assuming that the restriction *can* be made at the level of `argparse`, does it make more sense to do it in `argparse` or in the sub-command logic?

File: cli/tests/subcommands/test_dateparser.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    def setUp(self):
+        self.logger = logging.getLogger(__name__)
+
+    def test_parse(self):
+        time_zone = pytz.utc
+        now = datetime.datetime.now(time_zone)
+        with freeze_time(now):
+            self.assertEqual(now, dateparser.parse('now', time_zone))
+            self.assertEqual(now, dateparser.parse('NOW', time_zone))
```
Comment (DaoWen) [inline cli/tests/subcommands/test_dateparser.py]: `Now`?

File: cli/tests/subcommands/test_dateparser.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            self.assertEqual(now - datetime.timedelta(seconds=1), dateparser.parse('1 second ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(seconds=2), dateparser.parse('2 seconds ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(minutes=1), dateparser.parse('1 min ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(minutes=1), dateparser.parse('1 minute ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(minutes=2), dateparser.parse('2 minutes ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(hours=1), dateparser.parse('1 hour ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(hours=2), dateparser.parse('2 hours ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(days=1), dateparser.parse('1 day ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(days=2), dateparser.parse('2 days ago', time_zone))
+            self.assertEqual(now - datetime.timedelta(weeks=1), dateparser.parse('1 week ago', time_zone))
```
Comment (DaoWen) [inline cli/tests/subcommands/test_dateparser.py]: `1 WEEK AGO`?

File: cli/tests/subcommands/test_dateparser.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                             dateparser.parse('Fri Oct  6 16:37:59 CDT 2017', time_zone))
+            self.assertEqual(datetime.datetime(2003, 9, 25, 10, 49, 41,
+                                               tzinfo=tz.tzoffset(None, datetime.timedelta(hours=-3))),
+                             dateparser.parse('Thu, 25 Sep 2003 10:49:41 -0300', time_zone))
+            self.assertIsNone(dateparser.parse('1 day ago 1 hour ago', time_zone))
+            self.assertIsNone(dateparser.parse('1 day', time_zone))
+            self.assertIsNone(dateparser.parse('day ago', time_zone))
+            self.assertIsNone(dateparser.parse('day', time_zone))
+            self.assertIsNone(dateparser.parse('ago', time_zone))
+            self.assertIsNone(dateparser.parse('foo', time_zone))
```
Comment (DaoWen) [inline cli/tests/subcommands/test_dateparser.py]: What happens with an empty string?

File: cli/cook/subcommands/list.py (hunk 1 above)
Comment (DaoWen) [reply cli/cook/subcommands/list.py]: OK, it actually looks like nested groups aren't officially supported right now (see https://stackoverflow.com/a/43288171/1427124), so we're stuck with putting the logic elsewhere.

Comment (dposada) [reply cli/cook/subcommands/list.py]: Yeah, I tried nesting a group within the mutual exclusion group, and it didn't work :(

File: cli/cook/__main__.py [hunk 9]
```diff
@@ -21,5 +22,17 @@ def main(args=None):
         sys.exit(1)
 
 
+def sigint_handler(_, __):
+    """
+    Sets util.quit_running to True (which is read by other
+    threads to determine when to stop), and then exits.
+    """
+    util.quit_running = True
```
[the remark this answers is not in the document]
Comment (dposada) [reply cli/cook/__main__.py]: Agreed

File: cli/cook/dateparser.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    - add complexity we don't want
+    - slow them down
+    - make pyinstaller compatibility hard or impossible
+    """
+    date_time_string = date_time_string.strip()
+    date_time_string_lower = date_time_string.lower()
+    now = datetime.datetime.now(tz=time_zone)
+
+    if date_time_string_lower == 'now' or date_time_string_lower == 'today':
```
[the remark this answers is not in the document]
Comment (dposada) [reply cli/cook/dateparser.py]: Done

File: cli/tests/subcommands/test_dateparser.py (hunk 6 above)
Comment (dposada) [reply cli/tests/subcommands/test_dateparser.py]: Done

File: cli/tests/subcommands/test_dateparser.py (hunk 7 above)
Comment (dposada) [reply cli/tests/subcommands/test_dateparser.py]: Done ("3 WEEKS AGO")

File: cli/tests/subcommands/test_dateparser.py (hunk 8 above)
Comment (dposada) [reply cli/tests/subcommands/test_dateparser.py]: Added a unit test for empty string, and integration tests for e.g. `--submitted-after ""`

**Pull request merged**